bin/run-migration-job.py

```diff
@@ -0,0 +1,111 @@
+    - PGDATABASE: name of the database which will be used for migrations
+    - Any additional libpq-supported connection parameters
+          (https://www.postgresql.org/docs/current/libpq-envars.html)
+"""
+import argparse
+import asyncio
+import os
+import sys
+
+from buildpg import V, funcs, render
+
+from swoop.db import SwoopDB
+
+APPLICATION_ROLES = ["user_api", "user_caboose", "user_conductor"]
+
+
+def stderr(*args, **kwargs) -> None:
+    kwargs["file"] = sys.stderr
+    print(*args, **kwargs)
+
+
+def check_positive(value) -> int:
+    ivalue = int(value)
+    if ivalue <= 0:
+        raise argparse.ArgumentTypeError(
+            "%s is an invalid migration version number. Only \
+                positive values are supported."
+            % value
+        )
+    return ivalue
+
+
+async def run_migrations() -> None:
+    dbname: str = os.environ["PGDATABASE"]
+
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        "--action",
+        choices=["migrate", "rollback"],
+        help="An action to take, either migrate or rollback",
+    )
+    parser.add_argument(
+        "--version",
+        type=check_positive,
+        help="The migration version to which to migrate/rollback the database",
+    )
+
+    parser.add_argument(
+        "--wait_override",
+        choices=["true", "false"],
+        help="Override option to skip waiting for active connections to close. \
+        If specified, it is true, and if not specified it is false.",
+    )
+
+    args = parser.parse_args()
+
+    action = args.action
+    version = args.version
+    override = args.wait_override
+
+    swoop_db = SwoopDB()
+
+    if override == "true":
+        stderr(f"Applying migrate/rollback on database {dbname} to version {version}")
+        await swoop_db.migrate(target=version, database=dbname)
+    else:
+        async with swoop_db.get_db_connection(database="") as conn:
+            # Wait for all active connections from user roles to be closed
+            active = True
+            roles_list = V("usename") == funcs.any(APPLICATION_ROLES)
+            while active:
+                q, p = render(
+                    "SELECT * FROM pg_stat_activity WHERE datname = :db AND :un_clause",
+                    db=dbname,
+                    un_clause=roles_list,
+                )
+
+                records = await conn.fetch(q, *p)
+                if len(records) == 0:
+                    active = False
+
+            if action == "migrate":
+                stderr(f"Migrating database {dbname} to version {version}")
+            elif action == "rollback":
+                stderr(f"Rolling back database {dbname} to version {version}")
+            else:
+                stderr(
+                    f"Applying migrate/rollback on database {dbname} to \
+                    version {version}"
+                )
+            await swoop_db.migrate(target=version, database=dbname)
```

Comment:
You have different logic for applying the migrations between these two cases. It's all just logging, but my point remains: applying the migrations is not different whether you wait or not. It's the waiting that's different, an only that piece should be affected by the conditional.

Also, migrate vs rollback functionally changes how the `SwoopDB.migrate()` works. See how direction works in `https://github.com/Element84/dbami/blob/main/src/dbami/db.py#L413`. It's technically not a required parameter because the function can work out the path to a given target no matter the direction, but when it comes to ensuring the user's expectations are met with regard to only moving forward or backward the direction must be specified.

Lastly, the while loop is going to flood the cluster with queries, because it will run as fast as it possibly can. If you are waiting for things, it is good to ensure you are actually waiting in some way. In this case throwing in a sleep would suffice. 

Putting this together with the option changes I proposed, I would do this:

```suggestion
    async with swoop_db.get_db_connection() as conn:
        # Wait for all active connections from user roles to be closed
        active_sessions = not args.no_wait
        while active_sessions:
            active_sessions = await conn.fetchval(
                """
                SELECT EXISTS(
                    SELECT * FROM pg_stat_activity
                    WHERE
                        datname = current_database()
                        AND usename != current_user
                )
                """,
            )
            if active_sessions:
                time.sleep(2)
           
        if args.rollback:
            stderr(f"Rolling back database {dbname} to version {version}")
            direction = "down"
        else:
            stderr(f"Migrating database {dbname} to version {version}")
            direction = "up"

        await swoop_db.migrate(target=version, direction=direction, database=dbname)
```

Note that `SwoopDB.migrate()` establishes its own connection, but we probably don't want that. Several places in dbami we use the following pattern to only get a connection if one is not passed in:

```
async with AsyncExitStack() as stack:
    conn: asyncpg.Connection = kwargs.get(
        "conn",
    ) or await stack.enter_async_context(cls.get_db_connection(**kwargs))
```

We could reduce duplication and add this behavior to `migrate()` by putting this logic into the `get_db_connection()` method, maybe something like:

```
    @staticmethod
    @asynccontextmanager
    async def get_db_connection(**kwargs):
        conn = None
        try:
            if kwargs["conn"]:
                yield kwargs["conn"]
            else:
                conn = await asyncpg.connect(**kwargs)
                yield conn
        finally:
            if conn:
                await conn.close()
```

Then here you could call `await swoop_db.migrate(target=version, direction=direction, conn=conn)` to migrate with the reused connection.

Reply:
The `dbami` change is pending: https://github.com/Element84/dbami/pull/11. Once that is merged I can cut a release and this PR can use that change.

Reply:
I have released `dbami` v0.3.0, and that is now in the requirements.txt on main.

Reply:
Yeah, I was expecting a comment on this, as I saw that the migrate function in `dbami` established its own connection- good to know that we added logic to handle pre-existing connections.